File: NinePSharp.Messages.FSharp/messages/Twrite.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Twrite =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      Offset: uint64
      Count: uint32
      Data: byte[] }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Twrite
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Twrite span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.Offset); offset <- offset + 8
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Count); offset <- offset + 4
            if not (isNull this.Data) then
                this.Data.CopyTo(span.Slice(offset))

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let foffset = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let count = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let dataArr = data.Slice(offset, int count).ToArray()
        { Size = size; Tag = tag; Fid = fid; Offset = foffset; Count = count; Data = dataArr }

File: NinePSharp/messages/Twrite.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Twrite tag[2] fid[4] offset[8] count[4] data[count]
public readonly struct Twrite : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Twrite;
    public ushort Tag { get; }

    public uint Fid { get; }
    public ulong Offset { get; }
    public uint Count { get; }
    public ReadOnlyMemory<byte> Data { get; }

    public Twrite(ReadOnlyMemory<byte> data)
    {
        var span = data.Span;
        Size = BinaryPrimitives.ReadUInt32LittleEndian(span[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(span.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;

        Fid = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(offset, 4));
        offset += 4;

        Offset = BinaryPrimitives.ReadUInt64LittleEndian(span.Slice(offset, 8));
        offset += 8;

        Count = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(offset, 4));
        offset += 4;

        // Zero-copy: Reference the existing memory slice
        Data = data.Slice(offset, (int)Count);
    }
    
    public Twrite(uint size, ushort tag, uint fid, ulong offset, uint count, ReadOnlyMemory<byte> data)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        Offset = offset;
        Count = count;
        Data = data;
    }

    public Twrite(ushort tag, uint fid, ulong offset, ReadOnlyMemory<byte> data)
    {
        Tag = tag;
        Fid = fid;
        Offset = offset;
        Data = data;
        Count = (uint)data.Length;
        Size = (uint)(NinePConstants.HeaderSize + 4 + 8 + 4 + data.Length);
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Fid);
        offset += 4;

        BinaryPrimitives.WriteUInt64LittleEndian(data.Slice(offset, 8), Offset);
        offset += 8;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Count);
        offset += 4;

        Data.Span.CopyTo(data.Slice(offset, (int)Count));
    }
}
